Autoformat markdown files

sirreal opened the pull request:
## What?

- Autoformat markdown files as part of lint-staged.
- Apply autoformatting to markdown files in the repo.

## Why?

Autoformatting is used for JavaScript in Gutenberg. Why not for markdown as well? The consistency is nice.


## Testing Instructions

Confirm that the rendered markdown files are unchanged. The formatting only affects the source.

Try modifying some markdown and committing locally, the files should be formatted on save.

sirreal committed:
Format changelogs

sirreal committed:
Format markdown on lint-staged

sirreal committed:
Allow for arbitrary number of spaces matching in "- " list markdown

gziolo commented:
Great idea 👍🏻

sirreal committed:
DROPME: Add fake core merge file to test CI workflow

fabiankaegy commented:
Honestly not sure how I feel about it. 

I like the idea of standardized formatting for all the things. 

But I do worry that for markdown files this will make contributing even harder for non developers than it already is. 

A lot of the nice spelling fixes / documentation additions are coming from first time contributors / non code contributors. And it would be a shame to block all those PRs because some arbitrary formatting in MD files. 

Most contributions like that actually happen through the GitHub web UI where this formatting won't be auto applied. 

We already have this issue with the need for generating the manifest files after documentation changes. But I worry this will make it even harder. 

I would love to see us actually make that part of contributing easier by running those commands / auto formatting directly in CI if we can fix it automatically.

sirreal commented:
> But I do worry that for markdown files this will make contributing even harder for non developers than it already is.

That's a good concern. I don't _think_ this will change anything, it really doesn't matter if things are not all formatted on a given commit. As folks works with files, they'll tend towards the expected formatting.

I don't believe anything will block commits that aren't fully formatted, I've just pushed a change to test that so we'll see 🙂

sirreal commented:
Ok, there's a problem. A bunch of files are generated but are not formatted. Which causes CI to fail 😕  The generated files would need to be correctly formatted.

Related to your point 😀 

> I would love to see us actually make that part of contributing easier by running those commands / auto formatting directly in CI if we can fix it automatically.

I agree with you in general, _especially_ where these files are expected to have their contents be updated and committed. For formatting, it's not such a big deal if things aren't up to date.

sirreal committed:
Format as part of checking local changes

sirreal committed:
move markdown formatting to postdocs:build script

sirreal committed:
Run docs:build script

sirreal committed:
Add format:docs script and add it to lint-staged docs checking

sirreal committed:
npm run format

sirreal committed:
Autoformat everything

sirreal committed:
Anchor changelog regex to start/end of line

sirreal committed:
Update regex from #68049

sirreal commented on .github/workflows/check-backport-changelog.yml:
```diff
@@ -34,7 +34,7 @@ jobs:
                   # Find any changelog file that contains the Gutenberg PR link
                   gutenberg_pr_url="https://github\.com/WordPress/gutenberg/pull/${PR_NUMBER}"
-                  changelog_file=$(grep -rl "[-*] ${gutenberg_pr_url}" "${changelog_folder}" | head -n 1)
+                  changelog_file=$(grep -Erl "[-*] +${gutenberg_pr_url}" "${changelog_folder}" | head -n 1)
```
This is important to review. The backport changelog has changed its format to adhere to the standard markdown formatting. The regex would not correctly recognized the autoformatted markdown.

I've extracted this change to https://github.com/WordPress/gutenberg/pull/68049 so it can be landed separate from all the autoformatting work here.

sirreal commented:
> Ok, there's a problem. A bunch of files are generated but are not formatted. Which causes CI to fail 😕 The generated files would need to be correctly formatted.

I've addressed this by adding docs autoformatting in a new script `postdocs:build` and added that script to the docs lint-staged flow.

Unfortunately… it's quite slow 😞

sirreal committed:
Format components docs

sirreal committed:
format components docs
